contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,298 @@
+    /// time staking.
+    uint256 public boostedThreshold;
+
+    /// @notice The length of time a user receives boosted benefits
+    uint256 public boostedDuration;
+
+    mapping(address => UserInfo) public userInfos;
+
+    event Claim(address indexed user, uint256 amount);
+    event Deposit(address indexed user, uint256 amount);
+    event UpdateBaseGenerationRate(address indexed user, uint256 baseGenerationRate);
+    event UpdateBoostedDuration(address indexed user, uint256 boostedDuration);
+    event UpdateBoostedGenerationRate(address indexed user, uint256 boostedGenerationRate);
+    event UpdateBoostedThreshold(address indexed user, uint256 boostedThreshold);
+    event UpdateMaxCap(address indexed user, uint256 maxCap);
+    event Withdraw(address indexed user, uint256 amount);
+
+    /// @notice Initialize with needed parameters
+    /// @param _joe Address of the JOE token contract
+    /// @param _veJoe Address of the veJOE token contract
+    /// @param _baseGenerationRate Rate of veJOE generated per sec per JOE staked
+    /// @param _boostedGenerationRate Boosted rate of veJOE generated per sec per JOE staked
+    /// @param _boostedThreshold Percentage of total staked JOE user has to deposit to be boosted
+    /// @param _boostedDuration Length of time a user receives boosted benefits
+    function initialize(
+        IERC20Upgradeable _joe,
+        VeJoeToken _veJoe,
+        uint256 _baseGenerationRate,
+        uint256 _boostedGenerationRate,
+        uint256 _boostedThreshold,
+        uint256 _boostedDuration,
+        uint256 _maxCap
+    ) public initializer {
+        require(address(_joe) != address(0), "VeJoeStaking: unexpected zero address for _joe");
+        require(address(_veJoe) != address(0), "VeJoeStaking: unexpected zero address for _veJoe");
+        require(
+            _boostedGenerationRate > _baseGenerationRate,
+            "VeJoeStaking: expected _boostedGenerationRate to be greater than _baseGenerationRate"
+        );
+        require(_boostedThreshold <= 100, "VeJoeStaking: expected _boostedThreshold to be less than or equal to 100");
+        // TODO: Align on what the upper limit of maxCap should be
+        require(
+            _maxCap > 0 && _maxCap <= 100000,
+            "VeJoeStaking: expected new _maxCap to be greater than 0 and leq to 100000"
+        );
+
+        __Ownable_init();
+
+        maxCap = _maxCap;
+        joe = _joe;
+        veJoe = _veJoe;
+        baseGenerationRate = _baseGenerationRate;
+        boostedGenerationRate = _boostedGenerationRate;
+        boostedThreshold = _boostedThreshold;
+        // TODO: Align on what the upper limit of boostedDuration should be and add require check
+        boostedDuration = _boostedDuration;
+    }
+
+    /// @notice Set maxCap
+    /// @param _maxCap The new maxCap
+    function setMaxCap(uint256 _maxCap) external onlyOwner {
+        // TODO: Align on what the upper limit of maxCap should be
```

Comment:
I think we should add a validation here `require(_maxCap > maxCap)` this is to ensure no user has more veJOE than their maxCap would allow. If this were the case future changes to economic parameters could be thrown out by this. I have also asked MrBeaverTail and he agrees this number should never decrease.